atl-checker/src/algorithms/certain_zero/search_strategy/linear_constraints.rs

```diff
@@ -0,0 +1,565 @@
+        }
+    }
+
+    /// Visits an Expr and collects all valid linear terms.
+    /// Negated terms are handled using the sign parameter (either 1 or -1).
+    /// If this returns None, the expression is not linear.
+    fn collect_terms(&mut self, expr: &Expr, sign: i32) -> Option<()> {
+        match &expr.kind {
+            ExprKind::Number(n) => self.constant += sign * n,
+            ExprKind::OwnedIdent(ident) => {
+                // A variable with no explicit coefficient (which means the coefficient is 1)
+                if let Identifier::Resolved { owner, name } = ident.as_ref() {
+                    let coefficient = self.terms.entry(owner.symbol_id(name)).or_default();
+                    *coefficient += sign;
+                } else {
+                    panic!("Unresolved identifier")
+                }
+            }
+            ExprKind::UnaryOp(UnaryOpKind::Negation, expr) => {
+                // Unary negation simply flips the sign
+                self.collect_terms(expr, -sign)?;
+            }
+            ExprKind::BinaryOp(operator, lhs, rhs) => match operator {
+                BinaryOpKind::Addition => {
+                    self.collect_terms(lhs, sign)?;
+                    self.collect_terms(rhs, sign)?;
+                }
+                BinaryOpKind::Subtraction => {
+                    self.collect_terms(lhs, sign)?;
+                    self.collect_terms(rhs, -sign)?; // Note: flipped sign
+                }
+                BinaryOpKind::Multiplication => {
+                    // We found a potential term
+                    // One side must be a constant, the other must be a variable
+                    if let (ExprKind::Number(n), ExprKind::OwnedIdent(ident))
+                    | (ExprKind::OwnedIdent(ident), ExprKind::Number(n)) = (&lhs.kind, &rhs.kind)
+                    {
+                        if let Identifier::Resolved { owner, name } = ident.as_ref() {
+                            let coefficient = self.terms.entry(owner.symbol_id(name)).or_default();
+                            *coefficient += n * sign;
+                        } else {
+                            panic!("Unresolved identifier")
+                        }
+                    } else {
+                        return None;
+                    }
+                }
+                BinaryOpKind::Division => {
+                    // We found a potential term.
+                    // A variable divided with a constant is a linear expression,
+                    // but constant divided by variable is not.
+                    if let (ExprKind::OwnedIdent(ident), ExprKind::Number(n)) =
+                        (&lhs.kind, &rhs.kind)
+                    {
+                        if let Identifier::Resolved { owner, name } = ident.as_ref() {
+                            let coefficient = self.terms.entry(owner.symbol_id(name)).or_default();
+                            *coefficient += (1 / n) * sign;
```

Comment:
Wont the right side almost always become 0, as we are dealing with i32?

Reply:
It will. In general, using floats in the linear heuristic will give us more precision. We are going to need the `float-ord` crate to give a total ordering to floats (otherwise we can't do `max(float_dist_1, float_dist_2)`)

Reply:
Yes, I believe that I also ran into that at some point. How much of an impact do you think this will have on measuring the distance, using ints?

Reply:
Small, but some. It makes it more rough than it already is.